### ISSUE TITLE
Kotin lambda not recognised anymore as Function since Spring Cloud 2021.0.0 (Function 3.2.1)

### ISSUE THREAD

**Issue by awgroeneveld** [spring-cloud/spring-cloud-function#780]
**Describe the bug**
After upgrading to Spring Cloud Stream 2021.0.0 Kotlin Lambda's which in 2020.0.4 were interpreted as Functions are not correctly interpreted anymore.
After debugging, this is what I discovered and it seems like a changed order in Bean registration:

- in FunctionTypeUtils, we have the discoverFunctionType function consisting of the following code:
```
if (applicationContext.containsBean(functionName + FunctionRegistration.REGISTRATION_NAME_SUFFIX)) { // for Kotlin primarily
			FunctionRegistration fr = applicationContext
					.getBean(functionName + FunctionRegistration.REGISTRATION_NAME_SUFFIX, FunctionRegistration.class);
			return fr.getType().getType();
}
```
- this code searches for a bean with the suffix "_registration" for the function exposed as a Bean
- the problem now is that since commit 037f1b8bfe89c1a9e58dad36acca51ce1e43fb64 the KotlinLambdaToFunctionAutoConfiguration has changed it's implementation but it's method kotlinToFunctionTransformer is only called AFTER the discoverFunctionType call, therefore the bean is registered with the suffix too late !

**Sample**
Use Spring Cloud Stream Samples, Kotlin-rabbit-functions and upgrade to Spring Cloud 2021.0.0 and clearing the Spring Cloud Function versions.
Running the application results in the error "Type must be one of Supplier, Function or Consumer"
Debugging will display the above mentioned behaviour
Adding the name with postfix to the Bean like @Bean("javaFunction_registration") at least bypasses the check but likely fails functionally.
Will stay with 2.5.x with cloud 2020.0.4 for now

**Comment by olegz** [comment]
Sorry about that.

I was able to reproduce it, so with this fix I also updated the `kotlin-rabbit-functions` sample

**Comment by awgroeneveld** [comment]
Thanks a lot @olegz !

**Comment by mariusstaicu** [comment]
Hello, is there a way to manually go around this using 2021.0.0, without waiting for a new release ? 
Also, 3.2.2 is not released yet in maven.

**Comment by olegz** [comment]
3.1.6 is out and the fix was back-ported

**Comment by mariusstaicu** [comment]
is 3.1.6 included in spring cloud 2021.0.0 ? from what I see, 2021.0.0 works with 3.2.1
